File: src/algorithms/pathfinding/maze-generation/aldous-broder/sources/AldousBroder.java
import java.util.*;

// Aldous-Broder Maze — uniform random spanning tree via random walk
public class AldousBroder {
    public static int aldousBroder(int[][] grid, int[] start) {
        int rowCount = grid.length; // @step:initialize
        int colCount = grid[0].length; // @step:initialize
        boolean[][] visited = new boolean[rowCount][colCount]; // @step:initialize
        int passagesCarved = 0; // @step:initialize
        Random random = new Random();

        // Count total passage cells (odd row and odd col)
        int totalPassageCells = 0; // @step:initialize
        for (int rowIndex = 1; rowIndex < rowCount - 1; rowIndex += 2) {
            for (int colIndex = 1; colIndex < colCount - 1; colIndex += 2) {
                totalPassageCells++;
            }
        }

        int visitedCount = 0; // @step:initialize
        int currentRow = start[0]; // @step:initialize
        int currentCol = start[1]; // @step:initialize

        // Mark start as visited and carve it
        visited[currentRow][currentCol] = true; // @step:visit
        if (grid[currentRow][currentCol] == 1) {
            grid[currentRow][currentCol] = 0; // @step:carve-cell
            passagesCarved++;
        }
        visitedCount++;

        // Directions move 2 cells to passage-cell neighbors
        int[][] directions = {{-2, 0}, {2, 0}, {0, -2}, {0, 2}};

        // Iteration cap to prevent infinite loops
        int maxIterations = rowCount * colCount * 10;
        int iterations = 0;

        while (visitedCount < totalPassageCells && iterations < maxIterations) {
            iterations++;

            // Collect valid passage-cell neighbors
            List<int[]> validNeighbors = new ArrayList<>(); // @step:visit
            for (int[] direction : directions) {
                int neighborRow = currentRow + direction[0];
                int neighborCol = currentCol + direction[1];
                if (neighborRow < 1 || neighborRow >= rowCount - 1) continue;
                if (neighborCol < 1 || neighborCol >= colCount - 1) continue;
                validNeighbors.add(new int[]{neighborRow, neighborCol});
            }

            if (validNeighbors.isEmpty()) break;

            // Pick a random neighbor (random walk)
            int[] next = validNeighbors.get(random.nextInt(validNeighbors.size())); // @step:visit
            int nextRow = next[0];
            int nextCol = next[1];

            if (!visited[nextRow][nextCol]) {
                // Carve the wall between current and next
                int wallRow = currentRow + (nextRow - currentRow) / 2;
                int wallCol = currentCol + (nextCol - currentCol) / 2;
                grid[wallRow][wallCol] = 0; // @step:carve-cell
                passagesCarved++;

                // Carve the next passage cell
                if (grid[nextRow][nextCol] == 1) {
                    grid[nextRow][nextCol] = 0; // @step:carve-cell
                    passagesCarved++;
                }

                visited[nextRow][nextCol] = true; // @step:carve-cell
                visitedCount++;
            }

            currentRow = nextRow; // @step:visit
            currentCol = nextCol; // @step:visit
        }

        return passagesCarved; // @step:complete
    }
}

File: src/algorithms/pathfinding/maze-generation/aldous-broder/sources/aldous-broder.go
// Aldous-Broder Maze — uniform random spanning tree via random walk
package aldousbroder

import "math/rand"

type CellType int

const (
	CellEmpty CellType = iota
	CellWall
	CellStart
	CellEnd
)

type GridCell struct {
	Row      int
	Col      int
	CellType CellType
	State    string
}

type MazeResult struct {
	PassagesCarved int
}

func AldousBroder(grid [][]GridCell, startRow, startCol int) MazeResult {
	rowCount := len(grid) // @step:initialize
	colCount := 0
	if rowCount > 0 {
		colCount = len(grid[0]) // @step:initialize
	}
	visited := make([][]bool, rowCount)
	for rowIndex := range visited {
		visited[rowIndex] = make([]bool, colCount)
	} // @step:initialize
	passagesCarved := 0 // @step:initialize

	// Count total passage cells (odd row and odd col)
	totalPassageCells := 0 // @step:initialize
	for rowIndex := 1; rowIndex < rowCount-1; rowIndex += 2 {
		for colIndex := 1; colIndex < colCount-1; colIndex += 2 {
			totalPassageCells++
		}
	}

	visitedCount := 0        // @step:initialize
	currentRow := startRow   // @step:initialize
	currentCol := startCol   // @step:initialize

	// Mark start as visited and carve it
	visited[currentRow][currentCol] = true // @step:visit
	if grid[currentRow][currentCol].CellType == CellWall {
		grid[currentRow][currentCol].CellType = CellEmpty // @step:carve-cell
		passagesCarved++
	}
	visitedCount++

	directions := [][2]int{{-2, 0}, {2, 0}, {0, -2}, {0, 2}}
	maxIterations := rowCount * colCount * 10

	for iterations := 0; visitedCount < totalPassageCells && iterations < maxIterations; iterations++ {
		// Collect valid passage-cell neighbors
		type Position struct{ row, col int }
		var validNeighbors []Position // @step:visit
		for _, dir := range directions {
			neighborRow := currentRow + dir[0]
			neighborCol := currentCol + dir[1]
			if neighborRow < 1 || neighborRow >= rowCount-1 { continue }
			if neighborCol < 1 || neighborCol >= colCount-1 { continue }
			validNeighbors = append(validNeighbors, Position{neighborRow, neighborCol})
		}
		if len(validNeighbors) == 0 { break }

		// Pick a random neighbor (random walk)
		chosenIndex := rand.Intn(len(validNeighbors))
		nextRow := validNeighbors[chosenIndex].row // @step:visit
		nextCol := validNeighbors[chosenIndex].col // @step:visit

		if !visited[nextRow][nextCol] {
			// Carve the wall between current and next
			wallRow := currentRow + (nextRow-currentRow)/2
			wallCol := currentCol + (nextCol-currentCol)/2
			grid[wallRow][wallCol].CellType = CellEmpty // @step:carve-cell
			passagesCarved++

			if grid[nextRow][nextCol].CellType == CellWall {
				grid[nextRow][nextCol].CellType = CellEmpty // @step:carve-cell
				passagesCarved++
			}

			visited[nextRow][nextCol] = true // @step:carve-cell
			visitedCount++
		}

		currentRow = nextRow // @step:visit
		currentCol = nextCol // @step:visit
	}

	return MazeResult{PassagesCarved: passagesCarved} // @step:complete
}
